src/Bicep.Core/SourceCode/SourceArchive.cs

```diff
@@ -150,44 +187,127 @@ public static Stream PackSourcesIntoStream(Uri entrypointFileUri, IReadOnlyDicti
                 }
             }
 
             stream.Seek(0, SeekOrigin.Begin);
             return stream;
+        }
+
+        private static (string relativePath, string archivePath) CalculateRelativeAndArchivePaths(string path, string root, string rootName)
+        {
+            // Replace root of the path with root's "friendly" name to help avoid unintended
+            //   disclosure of user paths
+            var relativePath = Path.GetRelativePath(root, path);
+            Debug.Assert(!relativePath.StartsWith("../"), $"All paths should be under one of the roots");
+            relativePath = SourceCodePathHelper.NormalizeSlashes($"{rootName}{relativePath}");
+
+            // Handle illegal/problematic file characters in the path we use inside the archive
+            string archivePath = relativePath;
+            foreach (char ch in PathCharsToAvoid)
+            {
+                archivePath = archivePath.Replace(ch, '_');
+            }
+
+            // Place all sources files under "files/" in the archive
+            archivePath = Path.Join(FilesFolderName, archivePath);
 
-            (string location, string archivePath) CalculateFilePathsFromUri(Uri uri)
+            // Shorten if needed
+            archivePath = SourceCodePathHelper.Shorten(archivePath, MaxArchivePathLength);
+
+            return (SourceCodePathHelper.NormalizeSlashes(relativePath), SourceCodePathHelper.NormalizeSlashes(archivePath));
+        }
+
+        private static string GetPath(Uri uri)
+        {
+            return SourceCodePathHelper.NormalizeSlashes(uri.LocalPath);
+        }
+
+        private static string UniquifyArchivePath(IList<SourceFileInfoEntry> filesMetadata, string archivePath)
+        {
+            int suffix = 1;
+            string tryPath = archivePath;
+
+            while (filesMetadata.Any(f => PathHelper.PathComparer.Equals(f.ArchivePath, tryPath)))
+            {
+                suffix += 1;
+                tryPath = $"{archivePath}({suffix})";
+            }
+
+            return tryPath;
+        }
+
+        private static Dictionary<string, string> NameRoots(IDictionary<string, string> rootsMap, string entrypointRoot, string? cacheRoot)
+        {
+            var rootNamesMap = new Dictionary<string, string>();
+
+            // Entrypoint path is always as the top of the archive path hierarchy
+            rootNamesMap[entrypointRoot] = "";
+
+            // Remaining roots are "<root2>", "<root3>", etc., or <cache> for the module cache root
+            int i = 2;
+            foreach (var root in rootsMap.Values.Distinct().Where(r => r != entrypointRoot))
             {
-                var relativeLocation = CalculateRelativeFilePath(baseFolderUri, uri);
-                return (relativeLocation, relativeLocation);
+                if (PathHelper.PathComparer.Equals(root, cacheRoot))
+                {
+                    rootNamesMap[root] = "<cache>/";
```

Comment:
To play devil's advocate: should we actually be bundling modules from the cache? As a user, I wouldn't expect a given nuget package to bundle symbols from packages it depends on.

Some thoughts
- This creates possible leakage of "closed-source" data - for example, if a template spec (which has access managed via Azure RBAC) is published, this may give access to a wider audience.
- If the module publisher of a dependency wants to publish **their** source, they have the option to, and the user navigating the source code would be able to acces it.

Reply:
Please, play away!

What's in the cache is only the external module's compiled JSON, which is also compiled into the main.json, so as far as I know, this isn't exposing anything that isn't already being exposed (right?).  We likely will add an "--allow-republish" flag in the future where a module can say it's okay to also bundle its bicep sources into consumers' sources.  That's not implemented yet and would be opt-in only.

There is on the other hand a question as to how useful bundling the external module's JSON is in addition to its copy in the compiled main.json, as a matter of user interface.  I'm going to bring that up in a discussion meeting.

Reply:
> What we bundle from the cache is only the external module's compiled main.json, which is also compiled into the consumer's main.json. So as far as I know, this isn't exposing anything that isn't already being exposed (right?).

This is true for the Bicep registry, but not for template specs. To give an example for template specs, [this](https://github.com/Azure/bicep/blob/cbd44422001d53c21470c5e5e0f566fa8d38bab4/src/Bicep.Core.Samples/Files/baselines/Registry_LF/main.bicep#L57) is converted into [this syntax](https://github.com/Azure/bicep/blob/cbd44422001d53c21470c5e5e0f566fa8d38bab4/src/Bicep.Core.Samples/Files/baselines/Registry_LF/main.json#L379-L381) - rather than inlining the template, it's just referenced by resourceId.

> There is on the other hand a question as to how useful bundling the external module's JSON is in addition to its copy in the compiled main.json, as a matter of user interface. I'm going to bring that up in a discussion meeting.

Good point! If we could make navigating to and pulling down sources for a referenced module seamless, then this feels like it would address both concerns.